core/station/impl/src/services/system.rs

```diff
@@ -575,69 +662,257 @@ mod init_canister_sync_handlers {
+        }
+
+        Ok(())
+    }
+
+    pub async fn set_initial_permissions(
+        permissions: &[InitPermissionInput],
+    ) -> Result<(), ApiError> {
+        for permission in permissions {
+            let users = permission
+                .allow
+                .users
+                .iter()
+                .map(|id| HelperMapper::to_uuid(id.clone()).map(|uuid| *uuid.as_bytes()))
+                .collect::<Result<Vec<_>, _>>()?;
+
+            let user_groups = permission
+                .allow
+                .user_groups
+                .iter()
+                .map(|id| HelperMapper::to_uuid(id.clone()).map(|uuid| *uuid.as_bytes()))
+                .collect::<Result<Vec<_>, _>>()?;
+
+            let input = EditPermissionOperationInput {
+                resource: permission.resource.clone().into(),
+                auth_scope: Some(permission.allow.auth_scope.clone().into()),
+                users: Some(users),
+                user_groups: Some(user_groups),
+            };
+
+            PERMISSION_SERVICE.edit_permission(input)?;
+        }
+
+        Ok(())
+    }
+
+    fn specifier_has_reference_to_policy_id(
+        specifier: &RequestSpecifier,
+        policy_id: &UUID,
+    ) -> bool {
+        match specifier {
+            RequestSpecifier::EditRequestPolicy(resource_ids)
+            | RequestSpecifier::RemoveRequestPolicy(resource_ids) => match resource_ids {
+                ResourceIds::Any => false,
+                ResourceIds::Ids(ids) => ids.contains(policy_id),
+            },
+            _ => false,
+        }
+    }
+
+    pub fn set_initial_request_policies(
+        request_policies: &[InitRequestPolicyInput],
+    ) -> Result<(), ApiError> {
+        let mut add_request_policies = request_policies
+            .iter()
+            .map(|request_policy| {
+                let request_policy_id = request_policy
+                    .id
+                    .as_ref()
+                    .map(|id| HelperMapper::to_uuid(id.clone()).map(|uuid| *uuid.as_bytes()))
+                    .transpose();
+
+                let input = AddRequestPolicyOperationInput {
+                    specifier: request_policy.specifier.clone().into(),
+                    rule: request_policy.rule.clone().into(),
+                };
+
+                request_policy_id.map(|request_policy_id| (input, request_policy_id))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        // sorting criteria:
+        // - if a policy depends on another policy, the dependent policy should be added first
+        // - keep the original order of the policies otherwise
+        add_request_policies.sort_by(|a, b| {
+            if let Some(a_id) = &a.1 {
+                if specifier_has_reference_to_policy_id(&b.0.specifier, a_id) {
+                    return Ordering::Less;
+                }
+            }
+            Ordering::Greater
```

Comment:
I'm not sure this result leads to keeping the original order: shouldn't we compare the original indices here?

Reply:
I changed `Greater` to `Equal` which should preserve the creation order.